src/main/java/htsjdk/variant/variantcontext/writer/BCF2Writer.java

```diff
@@ -122,6 +122,12 @@
     private VCFHeader lastVCFHeaderOfUnparsedGenotypes = null;
     private boolean canPassOnUnparsedGenotypeDataForLastVCFHeader = false;
 
+    // is the header written to the output stream?
+    private boolean isHeaderWritten;
+
+    // is at least one part of body written to the output stream?
+    private boolean isBodyWritten;
+
```

Comment:
Nevermind @amilamanoj, it was just a proposal if you really want to be more verbose about the error and what happened. Just FYI, this was what I meant:

```java
// this integer indicates the status for writing the header down
// -1 indicates no problem to write the header
// 0 means header was already written
// 1 means body was already written
// otherwise, this is an unexpected error
private int shouldWriteHeader = -1;

@Override
public void setVCFHeader(final VCFHeader header) {
    if (shouldWriteHeader != -1) {
        final String msg;
        switch(shouldWriteHeader) {
            case 0: msg = "header already written"; break;
            case 1: msg = "body already written"; break;
            default: msg = "unexpected error code (" + shouldWriteHeader + ")";
        }
        throw new IllegalStateException("The header cannot be modified after the header or variants have been written to the output stream. Error was: " + msg);
    }

    ....

}
```